null_safety_examples/cookbook/forms/validation/lib/main.dart

```diff
@@ -0,0 +1,80 @@
+import 'package:flutter/material.dart';
+
+void main() => runApp(MyApp());
+
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    final appTitle = 'Form Validation Demo';
+
+    return MaterialApp(
+      title: appTitle,
+      home: Scaffold(
+        appBar: AppBar(
+          title: Text(appTitle),
+        ),
+        body: MyCustomForm(),
+      ),
+    );
+  }
+}
+
+// Create a Form widget.
+class MyCustomForm extends StatefulWidget {
+  @override
+  MyCustomFormState createState() {
+    return MyCustomFormState();
+  }
+}
+
+// Create a corresponding State class.
+// This class holds data related to the form.
+class MyCustomFormState extends State<MyCustomForm> {
+  // Create a global key that uniquely identifies the Form widget
+  // and allows validation of the form.
+  //
+  // Note: This is a GlobalKey<FormState>,
+  // not a GlobalKey<MyCustomFormState>.
+  final _formKey = GlobalKey<FormState>();
+
+  @override
+  Widget build(BuildContext context) {
+    // Build a Form widget using the _formKey created above.
+    return Form(
+      key: _formKey,
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: <Widget>[
+          // #docregion TextFormField
+          TextFormField(
+            // The validator receives the text that the user has entered.
+            validator: (value) {
+              if (value == null || value.isEmpty) {
+                return 'Please enter some text';
+              }
+              return null;
+            },
+          ),
+          // #enddocregion TextFormField
+          Padding(
+            padding: const EdgeInsets.symmetric(vertical: 16.0),
+            // #docregion ElevatedButton
+            child: ElevatedButton(
+              onPressed: () {
+                // Validate returns true if the form is valid, or false otherwise.
+                if (_formKey.currentState!.validate()) {
+                  // If the form is valid, display a snackbar. In the real world,
+                  // you'd often call a server or save the information in a database.
+                  ScaffoldMessenger.of(context)
+                      .showSnackBar(SnackBar(content: Text('Processing Data')));
+                }
```

Comment:
(nitpick) you could use type flow analysis instead of `!`:

```suggestion
              onPressed: () {
                var state = _formKey.currentState;
                // Validate returns true if the form is valid, or false
                // otherwise.
                if (state != null && state.validate()) {
                  // If the form is valid, display a snackbar. In the real
                  // world, you'd often call a server or save the information in
                  // a database.
                  ScaffoldMessenger.of(context)
                      .showSnackBar(SnackBar(content: Text('Processing Data')));
                }
```

Reply:
The reason I prefer the `!` instead of checking for null here is because a null `currentState` means that there is a bug in the code (e.g. the developer forgot to set the `_formKey` to the `Form` widget) and so the code should fail with an error rather than silently.

So I would write in production code is indeed `!`, the same way we didn't check for null `currentState` before null-safety.